pkg/master/controller.go

```diff
@@ -296,13 +302,67 @@ func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.Endpoint
 			}
 		}
 		if !contains {
-			return false, false
+			missingPorts = append(missingPorts, port)
 		}
 	}
+	if len(missingPorts) > 0 {
+		return false, false, missingPorts
+	}
 	for _, addr := range sub.Addresses {
 		if addr.IP == ip {
-			return true, len(sub.Addresses) == count
+			return true, len(sub.Addresses) == count, missingPorts
 		}
 	}
-	return true, false
+	return true, false, missingPorts
+}
+
+// * updateMasterServiceIfNeeded sets service attributes for the
+//     given apiserver service.
+// * updateMasterServiceIfNeeded expects that the service object it
+//     manages will be managed only by updateMasterServiceIfNeeded;
+//     therefore, to understand this, you need only understand the
+//     requirements and the body of this function.
+// * updateMasterServiceIfNeeded ensures that the correct ports are
+//     present on the service but will not stomp existing ports.
+//
+// Requirements:
+// * All apiservers MUST use updateMasterServiceIfNeeded and only
+//     updateMasterServiceIfNeeded to manage service attributes
+// * updateMasterService is called periodically from all apiservers.
+func updateMasterServiceIfNeeded(svc *api.Service, servicePorts []api.ServicePort, serviceType api.ServiceType) *api.Service {
+	// Determine if the service is in the format we expect
+	// (servicePorts are present and service type matches)
+	formatCorrect, missingPorts := checkServiceFormat(svc, servicePorts, serviceType)
+	if formatCorrect {
+		return nil
+	}
+	if len(missingPorts) > 0 {
+		// Missing ports were found. Existing ports are retained.
+		svc.Spec.Ports = append(svc.Spec.Ports, missingPorts...)
+	}
+	svc.Spec.Type = serviceType
+	return svc
+}
+
+// Determine if the service is in the correct format updateMasterServiceIfNeeded
+// expects (servicePorts are present and service type matches). Ports which are
+// missing are returned so that they can added.
+func checkServiceFormat(s *api.Service, ports []api.ServicePort, serviceType api.ServiceType) (formatCorrect bool, missingPorts []api.ServicePort) {
+	formatCorrect = true
+	for _, port := range ports {
+		contains := false
+		for _, servicePort := range s.Spec.Ports {
```

Comment:
If this is a partial match (port number matches but name doesn't, or vice-versa), `missingPorts` will be returned which overlap with existing ports. When you try to update the service, validation will fail on duplicate portNames or duplicate ports. Be sure to add tests for this case.

Limiting the mutations to setting the type and adding missing ports is better, but I still don't see how this prevents thrash between masters if a port name or number changes (e.g. we add the "dns" name to a port and an old master keeps trying to update the service to remove that name).